rfcs/plugin-api/RFC.md

```diff
@@ -0,0 +1,279 @@
+              },
+            }
+          }
+        }),
+      )
+    )
+}
+```
+
+❌ = dealbreaker
+
+|         Pros         |                       Cons                        |
+| :------------------: | :-----------------------------------------------: |
+|        Simple        | Every plugin takes entire core config as argument |
+| Backwards compatible |            Nesting becomes convoluted             |
+|                      |               no error handling ❌                |
+|                      |        unclear when to execute a plugin ❌        |
+
+## Proposal 3: `use` API
+
+```ts
+export async function setup({ defineConfig, env }) {
+	const pluginJson = await env.$import(
+		"https://cdn.jsdelivr.net/gh/samuelstroschein/inlang-plugin-json@1/dist/index.js",
+	)
+	const sdk = await env.$import("https://cdn.jsdelivr.net/gh/inlang/inlang@1/dist/sdk-js/index.js")
+
+	return defineConfig({
+		referenceLanguage: "en",
+	})
+		.use(
+			pluginJson({
+				pathPattern: "./languages/{language}.json",
+			}),
+		)
+		.use(
+			sdk({
+				languageNegotiation: {
+					strategies: [{ type: "localStorage" }, { type: "navigator" }],
+				},
+			}),
+		)
+}
+```
+
+|                       Pros                       |                      Cons                      |
+| :----------------------------------------------: | :--------------------------------------------: |
+| Apps can read, extend and modify previous config |    Host applications must pass defineConfig    |
+|            No deep nesting of plugins            |                breaking change                 |
+|                                                  | `use` seems unnecessary compared to proposal 4 |
+
+Have a setup function that includes a defineConfig function. The return of `defineConfig` can be extended by calling `use` on it.
+
+A major downside is two-fold increase in complexity:
+
+1. Host applications must pass `defineConfig` to plugins, complicating the "consumption" of the inlang.config.js file.
+2. Consumers face two different APIs to define the config: `defineConfig` for core stuff and `use` for plugin stuff. That's a bit implicit.
+3. Consumers need to understand the concept of `setup` and `defineConfig`. In the current design `defineConfig` is the only setup function and concept that needs to be understood.
+
+## Proposal 4: `plugins` property
+
+```ts
+export async function defineConfig(env) {
+	const pluginJson = await env.$import(
+		"https://cdn.jsdelivr.net/gh/samuelstroschein/inlang-plugin-json@1/dist/index.js",
+	)
+	const sdk = await env.$import("https://cdn.jsdelivr.net/gh/inlang/inlang@1/dist/sdk-js/index.js")
+
+	const standardLintRules = await env.$import(
+		"https://cdn.jsdelivr.net/gh/inlang/inlang@1/dist/sdk-js/index.js",
+	)
+
+	return {
+		referenceLanguage: "en",
+		plugins: [
+			// @ivanhofer @jannesblobel the "collection" discussion is resolved with this approach.
```

Comment:
I don't get why this would solve the "collection" problem

Reply:
A collection becomes a plugin. Thereby, the concept of a lint collection is not required anymore :)

Reply:
The collection has nothing to do with the core of inlang. It is something for `lint`-plugins. And there we still have the same problem

Reply:
The debate about lint collections was "is the concept of lint collections required [to define multiple lint rules at once]?". We decided to move lint collections out of core because we were in agreement that maybe not. With the introduction of plugins, the answer is a definitive no. A plugin can define multiple lint rules at once. 

After introducing the plugin API, the code in https://github.com/inlang/standard-lint-rules/blob/main/src/collection.ts is redundant. 

Before:

```ts
export async function defineConfig(env) {
  // import the rules you want to use
  const { standardLintRules } = await env.$import(
    "https://cdn.jsdelivr.net/gh/inlang/standard-lint-rules@1/dist/index.js"
  );

  return {
    referenceLanguage: "en",
    languages: ["en", "de", "fr"],
    lint: [
      standardLintRules() // <-- concept of a collection
    ],
  };
}
```

After:

```ts
export async function defineConfig(env) {
  // import the rules you want to use
  const { standardLintRules } = await env.$import(
    "https://cdn.jsdelivr.net/gh/inlang/standard-lint-rules@1/dist/index.js"
  );

  return {
    referenceLanguage: "en",
    plugins: [
      pluginJson(),
      standardLintRules() // <-- just a plugin
    ],
  };
}
```

Reply:
In my opinion nothing really changes. A lint plugin can already return an array of lint rules